core/src/main/scala/kafka/common/InterBrokerSendThread.scala

```diff
@@ -84,9 +74,130 @@ abstract class InterBrokerSendThread(name: String,
+            requestIterator.remove()
+          } else
+            pollTimeout = if (unsentRequests.hasRequests) AdminClient.DefaultReconnectBackoffMs else Long.MaxValue
+        }
+      }
+    }
+    networkClient.poll(pollTimeout, now)
+  }
+
+  private def checkDisconnects(now: Long): Unit = {
+    // any disconnects affecting requests that have already been transmitted will be handled
+    // by NetworkClient, so we just need to check whether connections for any of the unsent
+    // requests have been disconnected; if they have, then we complete the corresponding future
+    // and set the disconnect flag in the ClientResponse
+    val nodeIterator = unsentRequests.nodeIterator()
+    while (nodeIterator.hasNext) {
+      val node = nodeIterator.next
+      if (networkClient.connectionFailed(node)) {
+        // Remove entry before invoking request callback to avoid callbacks handling
+        // coordinator failures traversing the unsent list again.
+        val requests = unsentRequests.getRequests(node)
+        nodeIterator.remove()
+        for (request <- requests.asScala) {
+          val handler = request.callback
+          if (networkClient.authenticationException(node) == null)
+            error(s"Failed to send the following request due to authentication error: ${request.toString}")
+          handler.onComplete(new ClientResponse(request.makeHeader(request.requestBuilder().latestAllowedVersion()),
+            handler, request.destination, now /* createdTimeMs */ , now /* receivedTimeMs */ , true /* disconnected */ ,
+            null /* versionMismatch */ , null /* responseBody */))
+        }
+      }
+    }
+  }
 
+  private def failExpiredRequests(now: Long): Unit = {
+    // clear all expired unsent requests
+    val expiredRequests = unsentRequests.removeExpiredRequests(now, unsentExpiryMs)
+    for (request <- expiredRequests.asScala) {
+      debug(s"Failed to send the following request after $unsentExpiryMs ms: ${request.toString}")
+      val handler = request.callback
+      handler.onComplete(new ClientResponse(request.makeHeader(request.requestBuilder().latestAllowedVersion()),
+        handler, request.destination, now /* createdTimeMs */ , now /* receivedTimeMs */ , true /* disconnected */ ,
+        null /* versionMismatch */ , null /* responseBody */))
+    }
+  }
+
+  def wakeup(): Unit = networkClient.wakeup()
 }
 
 case class RequestAndCompletionHandler(destination: Node, request: AbstractRequest.Builder[_ <: AbstractRequest],
-                                       handler: RequestCompletionHandler)
\ No newline at end of file
+                                       handler: RequestCompletionHandler)
+
+class UnsentRequests {
+  private var unsent = new HashMap[Node, ArrayDeque[ClientRequest]]
+
+  def put(node: Node, request: ClientRequest): Unit = {
+    var requests = unsent.get(node)
+    if (requests == null) {
+      requests = new ArrayDeque[ClientRequest]
+      unsent.putIfAbsent(node, requests)
```

Comment:
nit: I think this could be a regular `put` since we don't have any concurrent callers

Reply:
+1